File: problem_solving/dp/py/1137_tribonacci_number.py
'''
https://leetcode.com/problems/n-th-tribonacci-number
similar to fibonacii..
t: O(N)
s: O(1)
'''

def tribonacci(n: int) -> int:
    if n == 0:
        return 0
    elif n == 1:
        return 1
    elif n == 2:
        return 1
    tn0 = 0
    tn1 = 1
    tn2 = 1
    tn3 = 0

    for i in range(3, n+1):
        tn3 = tn2 + tn1 + tn0
        tn0 = tn1
        tn1 = tn2
        tn2 = tn3
    return tn3



n = 4
#Output: 4
result = tribonacci(n)
print("result: " + str(result))

n = 25
#Output: 1389537
result = tribonacci(n)
print("result: " + str(result))

File: problem_solving/dp/cpp/1137_tribonacci_number.cpp
/*
'''
https://leetcode.com/problems/n-th-tribonacci-number
similar to fibonacii..
t: O(N)
s: O(1)
'''
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<algorithm>
#include<math.h>


int tri_bonacci(int n)  {
    if(n==0) {
        return 0;
    }
    if(n <= 2) {
        return 1;
    }
    int result = 0;
    int t0 = 0;
    int t1=1;
    int t2=1;
    for(int i = 3; i <= n; ++i) {
        result = t0 + t1 + t2;
        t0 = t1;
        t1= t2;
        t2= result;
    }
    return result;
}


int main() {
    int n = 4;
    //#Output: 4
    int result = tri_bonacci(n);
    std::cout << "result: " << result << std::endl;
    
    n = 25;
    //#Output: 1389537
    result = tri_bonacci(n);
    std::cout << "result: " << result << std::endl;
    
}
